src/librustdoc/html/render/print_item.rs

```diff
@@ -222,47 +221,83 @@ trait ItemTemplate<'a, 'cx: 'a>: askama::Template + fmt::Display {
+            fn item_and_mut_cx(&self) -> (&'a clean::Item, RefMut<'_, &'a mut Context<'cx>>) {
+                (&self.it, self.cx.borrow_mut())
+            }
+        }
 
-fn item_template_render_attributes_in_pre<'a: 'b, 'b, 'cx: 'a>(
-    templ: &'b impl ItemTemplate<'a, 'cx>,
-) -> impl fmt::Display + Captures<'a> + 'b + Captures<'cx> {
-    display_fn(move |f| {
-        let (item, cx) = templ.item_and_mut_cx();
-        let tcx = cx.tcx();
-        let v = render_attributes_in_pre(item, "", tcx);
-        write!(f, "{v}")
-    })
+        impl<'a, 'cx: 'a> $name<'a, 'cx> {
+            item_template_methods!($($methods)*);
+        }
+    };
 }
 
-fn item_template_render_assoc_items<'a: 'b, 'b, 'cx: 'a>(
-    templ: &'b impl ItemTemplate<'a, 'cx>,
-) -> impl fmt::Display + Captures<'a> + 'b + Captures<'cx> {
-    display_fn(move |f| {
-        let (item, mut cx) = templ.item_and_mut_cx();
-        let def_id = item.item_id.expect_def_id();
-        let v = render_assoc_items(*cx, item, def_id, AssocItemRender::All);
-        write!(f, "{v}")
-    })
+macro_rules! item_template_methods {
+    () => {};
+    (document $($rest:tt)*) => {
+        fn document<'b>(&'b self) -> impl fmt::Display + Captures<'a> + 'b + Captures<'cx> {
+            display_fn(move |f| {
+                let (item, mut cx) = self.item_and_mut_cx();
+                let v = document(*cx, item, None, HeadingOffset::H2);
+                write!(f, "{v}")
+            })
+        }
+        item_template_methods!($($rest)*);
+    };
+    (document_type_layout $($rest:tt)*) => {
+        fn document_type_layout<'b>(&'b self) -> impl fmt::Display + Captures<'a> + 'b + Captures<'cx> {
+            display_fn(move |f| {
+                let (item, cx) = self.item_and_mut_cx();
+                let def_id = item.item_id.expect_def_id();
+                let v = document_type_layout(*cx, def_id);
+                write!(f, "{v}")
+            })
+        }
+        item_template_methods!($($rest)*);
+    };
+    (render_attributes_in_pre $($rest:tt)*) => {
+        fn render_attributes_in_pre<'b>(&'b self) -> impl fmt::Display + Captures<'a> + 'b + Captures<'cx> {
+            display_fn(move |f| {
+                let (item, cx) = self.item_and_mut_cx();
+                let tcx = cx.tcx();
+                let v = render_attributes_in_pre(item, "", tcx);
+                write!(f, "{v}")
+            })
+        }
+        item_template_methods!($($rest:tt)*);
+    };
+    (render_assoc_items $($rest:tt)*) => {
+        fn render_assoc_items<'b>(&'b self) -> impl fmt::Display + Captures<'a> + 'b + Captures<'cx> {
+            display_fn(move |f| {
+                let (item, mut cx) = self.item_and_mut_cx();
+                let def_id = item.item_id.expect_def_id();
+                let v = render_assoc_items(*cx, item, def_id, AssocItemRender::All);
+                write!(f, "{v}")
+            })
+        }
+    };
+    ($method:tt $($rest:tt)*) => {
```

Comment:
```suggestion
    ($method:ident $($rest:tt)*) => {
```

And very likely needs to add another case with:

```rust
    ($token:tt $($rest:tt)*) => {
        compile_error!(concat!("unexpected token: ", stringify($token));
    }
```

Reply:
Since we only expect `ident`s for methods, seems like we can restrict them further by only allowing `ident`s as input.

In this case, we need to modify the match rules in both `item_template!` and `item_template_methods!`